src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs

```diff
@@ -777,9 +763,12 @@ private static void SubstituteConstraintTypes(ImmutableArray<TypeWithAnnotations
                 var refKind2 = param2.RefKind;
 
                 // Metadata signatures don't distinguish ref/out, but C# does - even when comparing metadata method signatures.
-                if (considerRefKindDifferences)
+                if ((refKindCompareMode & RefKindCompareMode.ConsiderDifferences) != 0)
                 {
-                    if (refKind1 != refKind2)
+                    if (refKind1 != refKind2 &&
+                        !((refKindCompareMode & RefKindCompareMode.AllowRefReadonlyVsInMismatch) != 0 &&
```

Comment:
This whole condition is hard to read. Consider extracting to a local function and breaking down the logic into statements.